### PR TITLE
Alex/implement whitelist for propose

### PR DESCRIPTION
**Pull request by asoong**

### CODE REVIEW

File: contracts/core/tokens/rebalancing-libraries/StandardProposeLibrary.sol [hunk 1]
```diff
@@ -99,6 +105,17 @@ library StandardProposeLibrary {
             "RebalancingSetToken.propose: Invalid or disabled proposed SetToken address"
         );
 
+        // Check proposed components on whitelist. This is to ensure managers are unable to add contract addresses
+        // to a propose that prohibit the set from carrying out an auction i.e. a token that only the manager possesses
+        IWhiteList rebalanceComponentWhitelist = IRebalancingSetFactory(_factory).rebalanceComponentWhitelist();
+        address[] memory components = ISetToken(_nextSet).getComponents();
+        for (uint256 i = 0; i < components.length; i++) {
```
Comment (felix2feng) [inline contracts/core/tokens/rebalancing-libraries/StandardProposeLibrary.sol]: We could reduce the calls to whitelist by passing the array of components to the whitelist contract

File: contracts/core/tokens/rebalancing-libraries/StandardProposeLibrary.sol [hunk 2]
```diff
@@ -99,6 +105,17 @@ library StandardProposeLibrary {
             "RebalancingSetToken.propose: Invalid or disabled proposed SetToken address"
         );
 
+        // Check proposed components on whitelist. This is to ensure managers are unable to add contract addresses
+        // to a propose that prohibit the set from carrying out an auction i.e. a token that only the manager possesses
+        IWhiteList rebalanceComponentWhitelist = IRebalancingSetFactory(_factory).rebalanceComponentWhitelist();
```
Comment (felix2feng) [inline contracts/core/tokens/rebalancing-libraries/StandardProposeLibrary.sol]: If we do not believe that the factory will change (which it shouldn't), we could `cache` the rebalanceComponentWhitelist in a state variable

Comment (asoong) [reply contracts/core/tokens/rebalancing-libraries/StandardProposeLibrary.sol]: +1

File: contracts/core/interfaces/IWhiteList.sol [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+ * The IWhiteList interface exposes the whitelist mapping to check components
+ */
+interface IWhiteList {
+
+    /* ============ External Functions ============ */
+
+    /**
+     * Validates address against white list
+     *
+     * @param  _address       Address to check
```
Comment (bweick) [inline contracts/core/interfaces/IWhiteList.sol]: return javadoc

Comment (asoong) [reply contracts/core/interfaces/IWhiteList.sol]: +1

File: contracts/core/interfaces/IWhiteList.sol [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        address _address
+    )
+        external
+        view
+        returns(bool);
+
+    /**
+     * Verifies an array of addresses against the whitelist
+     *
+     * @return address[]      Array of addresses to verify
```
Comment (bweick) [inline contracts/core/interfaces/IWhiteList.sol]: return vs param javadoc

Comment (asoong) [reply contracts/core/interfaces/IWhiteList.sol]: +1

File: contracts/lib/WhiteList.sol [hunk 5]
```diff
@@ -139,4 +139,25 @@ contract WhiteList is
     {
         return addresses;
     }
-}
\ No newline at end of file
+
+    /**
+     * Verifies an array of addresses against the whitelist
+     *
+     * @return address[]      Array of addresses to verify
```
Comment (bweick) [inline contracts/lib/WhiteList.sol]: should be param and add a return javadoc

Comment (felix2feng) [inline contracts/lib/WhiteList.sol]: params javadoc

Comment (asoong) [reply contracts/lib/WhiteList.sol]: +1

**Pull request merged**